File: 2021/08/04/MajorityElement.java
/**
 * https://leetcode.com/problems/majority-element
 * @author Andrew Jarombek
 * @since 8/4/2021
 */

import java.util.HashMap;

public class MajorityElement {
    public static int majorityElement(int[] nums) {
        int mode = -1;
        int mostOccurrences = 0;
        HashMap<Integer, Integer> occurrences = new HashMap<>();

        for (int num: nums) {
            Integer occurrence = occurrences.get(num);

            if (occurrence == null) {
                occurrence = 0;
            }

            if (occurrence + 1 > mostOccurrences) {
                mode = num;
                mostOccurrences = occurrence + 1;
            }

            occurrences.put(num, occurrence + 1);
        }

        return mode;
    }

    public static void main(String[] args) {
        assert majorityElement(new int[] {3, 2, 3}) == 3;
    }
}

File: 2023/03/24/cpp/majorityElement.cpp
// https://leetcode.com/problems/majority-element/
// Author: Andrew Jarombek
// Date: 3/24/2023

#include <cassert>
#include <vector>
#include <unordered_map>

using namespace std;

int majorityElement(vector<int>& nums) {
    unordered_map<int, int> map {};

    for (int i = 0; i < nums.size(); i++) {
        int num = nums[i];
        int value = map[num] + 1;
        map[num] = value;

        if (value > nums.size() / 2) {
            return num;
        }
    }

    return -1;
}

int main() {
    vector<int> nums {3, 2, 3};
    assert(majorityElement(nums) == 3);
}